free-3.md

```diff
@@ -0,0 +1,230 @@
+and they should still be able to use _most_ of those within the tree framework.
+
+### Related work
+
+I don't know of any collected efforts to tackle this problem all at once. That
+being said there are some cool bits of code floating around the web that might
+serve as good inspiration.
+
+The first of these is some code which prints out binary trees in ASCII. You can
+find it
+[here](https://datastructuresblog.wordpress.com/2007/12/21/printing-binary-trees-in-ascii/)
+. It produces output like:
+
+```
+     4
+    / \
+   2   6
+  /   / \
+ 1   5   7
+```
+
+It seems like this would be a nice feature to have in any command line
+interface for exploring a tree.
+
+Another thing that comes to mind is the visual binary tree explorer written
+originally by Arsen Gogeshvili and enhanced by Prof. O'Neill, found
+[here](https://www.cs.hmc.edu/~oneill/BST-Applet/avltree.html). It allows you
+look took at how Red Black and AVL trees work, step by step. One could imagine
+the visual debugging/exploration interface looking a lot like this.
+
+## The Project
+This section examines whether the idea makes for a good CS 111 project.
+
+
+### Suitability
+
+I think this would actually a reasonable project for DSLs. Most of the time
+would be spent trying to figure out how to fit all the tree ideas (in
+particular metadata, metadata-yielding procedures, and meta-data dependencies
+inside the host language. That being said, implementation could also end up
+sucking up a lot of time, so it is hard to say for sure.
+
+I think extensions like the visual and/or command line explorer would be pretty
+fun features, but may not be as important or interesting from a DSL perspective
+(although the command line explorer would need its own control DSL!).
+
+### Scope
+
+I think this project would be well scoped, partly because of the scope of the
+features I've mentioned so far, but also because it has a variable scope. It
+would be really easy to say, "Hey , that CLI for tree exploring was a cool
+idea, but it's just too much work - we won't do it." Similarly, it'd be easy to
+say, "Hey, that CLI explorer control system seems like a good opportunity for a
+DSL - let's do that after all because we have the extra time."
+
+### Benefits and drawbacks
+
+I think this is a pretty cool opportunity to explore an internal DSL and think
+about how to weave the internal DSL and the host language seemlessly together.
+I also think that the metadata dependency problem is sort of cool and it would
+be fun to design a DSL which allows the users to express the necessary input to
+that problem in a way that feels natural.
+
+I also think that the variable scope would be a really excellent perk from a
+class project perspective - it would be easy to scale the project to the amount
+of time implementation ends up taking.
+
+Unfortunately I think that in the end of the day not too many people actually
+care about this problem. Think about it, who writes different types of trees?
```

Comment:
I'm probably one of those people who doesn't really care (sorry).  I guess in general I'm not that excited about trees.